### DISCUSSION TITLE
Why is there no implementation of Responder for i64?

### DISCUSSION THREAD

**Discussion by J-Cake** [rwf2/Rocket#2677] [Questions] [answered]
Hi frens, I'm struggling to work around an issue involving a `Responder`. I would like to build a route which when errors occur, provides as much information as possible. For fun, let's say the Route is defined like this: 
```Rust
#[derive(Default, Debug, Clone, Serialize, Deserialize, Responder)]
pub enum Error {
    UserNotFound {
        user: UserId
    },

    OperationNotPermitted {
        user: UserId,
        scope: Scope
    }

    // ---
}

#[derive(Default, Debug, Clone, Serialize, Deserialize, Responder)]
pub enum UserId {
    PrimaryKey(i64),
               ^^^ the trait `Responder<'_, '_>` is not implemented for `i64`
    Mail(String),
    LDAPName(String),
    LDAPOrMail(String)
}

#[post(/api)]
pub async fn do_something() -> Result<Json<SomeSuccessValue>, Error> {
    // ---
}
```

However, the error I'm getting is that *the trait `Responder<'_, '_>` is not implemented for `i64`*. My question here is twofold: 
 1. Why is it not implemented for most standard primitives? 
 2. How can I implement it? Keeping in mind that Rust won't allow the implementation of 3rd-party traits on 3rd-party types. 

Thankss

**Comment by SergioBenitez** [comment, marked as answer]
> 1. Why is it not implemented for most standard primitives?

It's not implemented because there's no obviously correct, semantically equivalent HTTP response to the primitives. In other words, it's not obvious what the correct HTTP response for an integer should be. Should it be the string representation of the integer? If so, what is the content-type, and in which base? Should it be a status code? Since it's not obvious, we can't implement `Responder`.

> 2. How can I implement it? Keeping in mind that Rust won't allow the implementation of 3rd-party traits on 3rd-party types.

You cannot. How you go around this depends and what you want the responder to do when it is a `PrimaryKey`. If you want it to return a string representation, then convert the `i64` into a string before placing it into your `PrimaryKey`. Or implement `Responder` directly:

```rust
impl<'r> Responder<'r, 'static> for UserId {
    fn respond_to(self, req: &'r Request<'_>) -> response::Result<'static> {
        let string = match self {
            UserId::PrimaryKey(n) => n.to_string(),
            UserId::Mail(s) | UserId::LDAPName(s) | UserId::LDAPOrMail(s) => s,
        };

        string.respond_to(req)
    }
}
```

**Reply by J-Cake** [reply]
This makes sense. The solution ended up being far more context-dependent than I initially thought. I'm working with `async-graphql`, so it turns out that in my case, implementing `Scalar` on `UserId` was exactly what I needed. 

As for the error type, well I just serialised it into JSON manually 😆 

Thanks for your help
